spec/datadog/profiling/collectors/discrete_dynamic_sampler_spec.rb

```diff
@@ -0,0 +1,177 @@
+        # With each sample taking 0.01 seconds, we can afford to do 2 of these every second.
+        # At an event rate of 100/sec we can sample 2% of total events.
+        stats = simulate_load(duration_seconds: 60, events_per_second: 100, sampling_seconds: 0.01)
+        expect(sampler.probability).to be_between(1, 3)
+        expect(stats[:sampling_ratio]).to be_between(0.01, 0.03)
+        expect(stats[:overhead]).to be < max_overhead_target
+      end
+    end
+  end
+
+  context 'when under a variable load' do
+    context 'containing lots of short spikes' do
+      it 'will readjust to decrease sampling rate' do
+        # Baseline
+        simulate_load(duration_seconds: 10, events_per_second: 10, sampling_seconds: 0.01)
+        p_baseline = sampler.probability
+
+        # We'll spike every 5 seconds. Sampler should have some short term memory so
+        # after a spiking period it should be using a lower probability
+        simulate_load(duration_seconds: 1, events_per_second: 1000, sampling_seconds: 0.01)
+        simulate_load(duration_seconds: 5, events_per_second: 10, sampling_seconds: 0.01)
+        simulate_load(duration_seconds: 1, events_per_second: 1000, sampling_seconds: 0.01)
+        simulate_load(duration_seconds: 5, events_per_second: 10, sampling_seconds: 0.01)
+        simulate_load(duration_seconds: 1, events_per_second: 1000, sampling_seconds: 0.01)
+        simulate_load(duration_seconds: 10, events_per_second: 10, sampling_seconds: 0.01)
+
+        p_after_spikes = sampler.probability
+
+        expect(p_after_spikes).to be < p_baseline
+      end
+    end
+
+    context 'with a big spike at the beginning' do
+      it "won't wait until the next sample to adjust" do
+        # We'll start with a very big load at the beginning. This should move the sampler towards
+        # having very low sampling probabilities (i.e. a big sampling interval). We want to validate
+        # that the sampler can readjust in-between samples so it can react to load pattern changes
+        # that may happen inbetween. If that weren't the case, after we go down to a very low event
+        # baseline, entire minutes could pass before we even decide to sample again and realize that
+        # we've been sampling nothing for a while.
+        simulate_load(duration_seconds: 5, events_per_second: 100000, sampling_seconds: 0.01)
+        p1 = sampler.probability
+        expect(p1).to be < 0.1 # %
+
+        # With such a low probability, our sampling skip is >1000 so if we relied on samples alone
+        # for adjustment, the 10 events generated in the following 2 loads would not trigger this
+        # readjustment.
+
+        simulate_load(duration_seconds: 5, events_per_second: 1, sampling_seconds: 0.01)
+        p2 = sampler.probability
+        expect(p2).to be > p1
+
+        simulate_load(duration_seconds: 5, events_per_second: 1, sampling_seconds: 0.01)
+        p3 = sampler.probability
+        expect(p3).to be > p2
```

Comment:
I wonder if we should be a bit stricter in our tests here, e.g. test that at some point we do see `num_samples` > 0 after an adjustment period.